### PR TITLE
fix initialization of QGLWidget waveform widgets & renderers 

### PR DESCRIPTION
**Pull request by Be-ing**
QGLWidget documentation says:
> The initialization of OpenGL rendering state, etc. should be done by overriding
> the initializeGL() function, rather than in the constructor of your QGLWidget
> subclass.

Calling QOpenGLFunctions_2_1::initializeOpenGLFunctions in the QGLWidget
constructors crashed on macOS:
https://bugs.launchpad.net/mixxx/+bug/1877487
I am surprised it did not crash on Linux and Windows.

The calls to makeCurrent in the QGLWidget constructors were futile; that
does nothing before Qt calls QGLWidget::initializeGL. This can be verified
by adding:
```
qDebug() << "Context current" << (QGLContext::currentContext() == context());
```
after the call to makeCurrent.

### CODE REVIEW

File: src/waveform/waveformwidgetfactory.cpp [hunk 1]
```diff
@@ -530,7 +530,7 @@ bool WaveformWidgetFactory::setWidgetTypeFromHandle(int handleIndex, bool force)
         // it since the size didn't change.
         //viewer->resize(viewer->size());
         widget->resize(viewer->width(), viewer->height());
-        widget->setTrack(pTrack);
+        //         widget->setTrack(pTrack);
```
Comment (Be-ing) [inline src/waveform/waveformwidgetfactory.cpp:533]: QGLWidget::initializeGL needs to be called on `widget->getWidget()` before `widget->setTrack(pTrack)`. The call to `widget->resize` above causes Qt to call QGLWidget::initializeGL, however, I could not figure out how to make Qt process that signal before the call to `widget->setTrack(pTrack)` here. I tried `QApplication::QApplication::sendPostedEvents(widget->getWidget())` and `QApplication::processEvents` but neither worked. Does anyone have ideas how to solve this?

**Comment by Be-ing** [comment]
This bug was introduced by #2094.

**Comment by Be-ing** [comment]
QOpenGLWidget has the [same API](https://doc.qt.io/qt-5/qopenglwidget.html#resource-initialization-and-cleanup) so this is forward compatible:
> The QOpenGLWidget's associated OpenGL context is guaranteed to be current whenever initializeGL() and paintGL() are invoked. Do not attempt to create OpenGL resources before initializeGL() is called. For example, attempting to compile shaders, initialize vertex buffer objects or upload texture data will fail when done in a subclass's constructor. These operations must be deferred to initializeGL().

File: src/waveform/waveformwidgetfactory.cpp [hunk 2]
```diff
@@ -530,9 +531,10 @@ bool WaveformWidgetFactory::setWidgetTypeFromHandle(int handleIndex, bool force)
         // it since the size didn't change.
         //viewer->resize(viewer->size());
         widget->resize(viewer->width(), viewer->height());
-        widget->setTrack(pTrack);
         widget->getWidget()->show();
         viewer->update();
+        QApplication::sendPostedEvents(widget->getWidget());
+        //         widget->setTrack(pTrack);
```
Comment (Be-ing) [inline src/waveform/waveformwidgetfactory.cpp:537]: QGLWidget::initializeGL needs to be called on `widget->getWidget()` before `widget->setTrack(pTrack)`. The call to `widget->resize` above causes Qt to call QGLWidget::initializeGL, however, I could not figure out how to make Qt process that signal before the call to `widget->setTrack(pTrack)` here. I tried `QApplication::sendPostedEvents(widget->getWidget())` and `QApplication::processEvents` but neither worked. Does anyone have ideas how to solve this?

**Comment by Be-ing** [comment]
😩 
```
Date/Time:             2020-12-06 06:13:40.068 -0600
OS Version:            Mac OS X 10.13.1 (17B1003)
Report Version:        12
Anonymous UUID:        1A32B552-30C0-2E9F-681B-B995D82E7AC4

Sleep/Wake UUID:       8FA21C8F-726C-4B01-BF4E-4FD951969E22

Time Awake Since Boot: 910000 seconds
Time Since Wake:       110000 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x00000000000000d8
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [0]

VM Regions Near 0xd8:
--> 
    __TEXT                 000000010a1ea000-000000010aa16000 [ 8368K] r-x/r-x SM=COW  /Users/USER/Desktop/Mixxx.app/Contents/MacOS/mixxx

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   org.mixxx.mixxx               	0x000000010a6dbeee GLSLWaveformRendererSignal::loadTexture() + 126
1   org.mixxx.mixxx               	0x000000010a7f9409 BaseTrackPlayer::loadingTrack(std::__1::shared_ptr<Track>, std::__1::shared_ptr<Track>) + 57
```

**Comment by Be-ing** [comment]
```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x8)
  * frame #0: 0x0000000101ed466b QtGui`QOpenGLContext::isOpenGLES() const + 11
    frame #1: 0x00000001010cc248 QtOpenGL`QGLFramebufferObject::QGLFramebufferObject(int, int, unsigned int) + 88
    frame #2: 0x00000001004f24e7 mixxx`GLSLWaveformRendererSignal::createFrameBuffers() + 87
    frame #3: 0x000000010050b3de mixxx`WaveformWidgetRenderer::resize(int, int, float) + 94
    frame #4: 0x000000010051d3ed mixxx`WaveformWidgetAbstract::resize(int, int) + 109
    frame #5: 0x0000000100517c39 mixxx`WaveformWidgetFactory::setWidgetTypeFromHandle(int, bool) + 569
    frame #6: 0x0000000102ab51ce QtCore`QMetaObject::activate(QObject*, int, int, void**) + 1934
    frame #7: 0x000000010178846d QtWidgets`QComboBoxPrivate::_q_emitCurrentIndexChanged(QModelIndex const&) + 189
    frame #8: 0x000000010178ce7f QtWidgets`QComboBoxPrivate::setCurrentIndex(QModelIndex const&) + 607
    frame #9: 0x000000010178cb94 QtWidgets`QComboBoxPrivate::_q_itemSelected(QModelIndex const&) + 84
    frame #10: 0x0000000102ab52a9 QtCore`QMetaObject::activate(QObject*, int, int, void**) + 2153
    frame #11: 0x000000010178a9d9 QtWidgets`QComboBoxPrivateContainer::eventFilter(QObject*, QEvent*) + 1193
    frame #12: 0x0000000102a86a11 QtCore`QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*) + 241
    frame #13: 0x000000010168fb04 QtWidgets`QApplicationPrivate::notify_helper(QObject*, QEvent*) + 244
    frame #14: 0x00000001016929b2 QtWidgets`QApplication::notify(QObject*, QEvent*) + 7506
    frame #15: 0x0000000100372346 mixxx`MixxxApplication::notify(QObject*, QEvent*) + 310
    frame #16: 0x0000000102a86746 QtCore`QCoreApplication::notifyInternal2(QObject*, QEvent*) + 166
    frame #17: 0x0000000101690422 QtWidgets`QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool, bool) + 898
    frame #18: 0x00000001016e9078 QtWidgets`QWidgetWindow::handleMouseEvent(QMouseEvent*) + 2376
    frame #19: 0x00000001016e7a9e QtWidgets`QWidgetWindow::event(QEvent*) + 238
    frame #20: 0x000000010168fb19 QtWidgets`QApplicationPrivate::notify_helper(QObject*, QEvent*) + 265
    frame #21: 0x0000000101690e32 QtWidgets`QApplication::notify(QObject*, QEvent*) + 466
    frame #22: 0x0000000100372346 mixxx`MixxxApplication::notify(QObject*, QEvent*) + 310
    frame #23: 0x0000000102a86746 QtCore`QCoreApplication::notifyInternal2(QObject*, QEvent*) + 166
    frame #24: 0x0000000101e91597 QtGui`QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) + 3751
    frame #25: 0x0000000101e78abb QtGui`QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) + 203
    frame #26: 0x0000000108138518 libqcocoa.dylib`QCocoaEventDispatcherPrivate::postedEventsSourceCallback(void*) + 40
    frame #27: 0x00007fff4837aad1 CoreFoundation`__CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
    frame #28: 0x00007fff4843254c CoreFoundation`__CFRunLoopDoSource0 + 108
    frame #29: 0x00007fff4835dac0 CoreFoundation`__CFRunLoopDoSources0 + 208
    frame #30: 0x00007fff4835cf3d CoreFoundation`__CFRunLoopRun + 1293
    frame #31: 0x00007fff4835c797 CoreFoundation`CFRunLoopRunSpecific + 487
    frame #32: 0x00007fff47671866 HIToolbox`RunCurrentEventLoopInMode + 286
    frame #33: 0x00007fff476714df HIToolbox`ReceiveNextEventCommon + 366
    frame #34: 0x00007fff47671354 HIToolbox`_BlockUntilNextEventMatchingListInModeWithFilter + 64
    frame #35: 0x00007fff4596ea23 AppKit`_DPSNextEvent + 2085
    frame #36: 0x00007fff46103e6c AppKit`-[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 3044
    frame #37: 0x00007fff45963831 AppKit`-[NSApplication run] + 764
    frame #38: 0x00000001081373cd libqcocoa.dylib`QCocoaEventDispatcher::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) + 2669
    frame #39: 0x0000000102a82167 QtCore`QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) + 487
    frame #40: 0x0000000102a86da2 QtCore`QCoreApplication::exec() + 130
    frame #41: 0x000000010000d4ef mixxx`main + 1775
    frame #42: 0x00007fff6f9f3145 libdyld.dylib`start + 1
    frame #43: 0x00007fff6f9f3145 libdyld.dylib`start + 1
```

**Comment by Be-ing** [comment]
It is only the GLSL waveforms that are crashing with this branch on macOS. The other OpenGL waveforms work, although performance is terrible.

**Comment by uklotzde** [comment]
Good catch! Looks like we are relying on undefined behavior that varies between platforms and Qt version :(

**Comment by JoergAtGithub** [comment]
I just checked the latest build on Windows7. Initially everything works, but if I open the Preference dialog the waveforms disapear (just black area), while the spinnies continue to rotate.

**Comment by Be-ing** [comment]
That is expected with the latest WIP commit. What I'm interested in is whether Mixxx crashes switching between waveform types (without the spinnies showing)?

**Comment by JoergAtGithub** [comment]
Just tested this:
1.) Initially waveforms and spinnies worked
2.) I disabled the spinnies in the settings -> Waveforms still work
3.) I opened the preference dialog -> Waveforms disapear immediately
4.) Switching waveform types doesn't show any effect
5.) Than I tried changing the Waveform Overview Type -> Mixxx crashed

File: src/waveform/widgets/waveformwidgetabstract.h [hunk 3]
```diff
@@ -43,4 +44,22 @@ class WaveformWidgetAbstract : public WaveformWidgetRenderer {
...
+/// GLWaveformWidgetAbstract is a WaveformWidgetAbstract & QGLWidget that has
+/// a GLWaveformRenderer member which requires initialization that must be
+/// deferred until Qt calls QGLWidget::initializeGL and cannot be done in the
+/// constructor.
+class GLWaveformWidgetAbstract : public WaveformWidgetAbstract, public QGLWidget {
+  public:
+    GLWaveformWidgetAbstract(const QString& group, QWidget* parent);
+
+  protected:
+    void initializeGL() override {
```
Comment (Be-ing) [inline src/waveform/widgets/waveformwidgetabstract.h:56]: Wtf this is getting called on Linux but not macOS....

Comment (JoergAtGithub) [reply src/waveform/widgets/waveformwidgetabstract.h:56]: Did you tried a clean rebuild?

**Comment by Be-ing** [comment]
I just pushed a proof-of-concept with QOpenGLWidget. I could not figure out how to get QGLWidget to call `initializeGL` on macOS. I think QGLWidget is hopelessly broken with Qt5 on macOS.

With this proof-of-concept, the OpenGL waveform widgets are usually black and occasionally flicker to show the waveform. UI responsiveness is good if no level meters are showing but becomes unusable with meters showing.

**Comment by Be-ing** [comment]
> I could not figure out how to get QGLWidget to call initializeGL on macOS. 

I hacked around this by initializing OpenGL stuff from the renderers' `draw` functions. Ugly, but it works.

**Comment by Be-ing** [comment]
Tested on macOS and Linux. Testing on Windows again would be appreciated.

**Comment by JoergAtGithub** [comment]
This works very good under Windows7. With Angle disabled and my Intel OpenGL driver I see much higher frame rates (50-60fps) and less CPU load that before. Only the the Qt GL waveforms are as slow as before. With Angle enabled, the frame rate is also slightly better.

**Comment by Be-ing** [comment]
Interesting. Thank you for testing. So even though it wasn't crashing on Windows, OpenGL wasn't actually working right.

**Comment by JoergAtGithub** [comment]
I just noticed, that the waveform type changes after each skin change to RGB (GL)

**Comment by Be-ing** [comment]
Whoa, the Wayland Qt platform plugin works now!

**Comment by Be-ing** [comment]
This branch shouldn't affect the preferences at all. Is that bug also present with the 2.3 branch?

**Comment by JoergAtGithub** [comment]
Just checked, also plain 2.3 changes waveform type to RGB (GL) on skin change, therefore it's not related to this PR.

**Comment by Be-ing** [comment]
This has no notable performance impact on Linux as far as I can tell.

**Comment by Holzhaus** [comment]
> This has no notable performance impact on Linux as far as I can tell.

The issue is that OpenGL hides a lot of magic on the OpenGL drivers. On Linux, we should verify this with both Mesa and Nvidia's proprietary drivers. What are you using?

On windows, testing with Intel, AMD and Nvidia would be ideal, but I guess we have to take what we can get considering that we don't have an abundance of windows testers.

**Comment by Be-ing** [comment]
I don't think testing this PR on a wide range of OSes and GPUs before merging is feasible. This is a critical bug fix. We should merge it ASAP. AFAIK there is no reason this should cause any regressions. It may have more performance benefits with other OS/GPU setups too, but we probably won't know if we leave it unmerged.

**Comment by Holzhaus** [comment]
I don't want to hold this back artificially. You already tested macOS and Linux (what GPU vendor?). Apparently It works fine on Windows/Intel, too. I can test on Mesa (Gallium/AMD). If someone from our dev team had an Nvidia card (AFAIK their Linux driver shared a lot of code with the Windows driver), we're golden.

**Comment by Be-ing** [comment]
Both my Linux laptops have integrated Intel GPUs. This old Macbook does too.

File: src/waveform/renderers/glslwaveformrenderersignal.cpp [hunk 4]
```diff
@@ -238,19 +237,23 @@ void GLSLWaveformRendererSignal::onSetTrack() {
...
+        return;
+    }
     loadTexture();
 }
 
 void GLSLWaveformRendererSignal::draw(QPainter* painter, QPaintEvent* /*event*/) {
-    if (!m_framebuffer || !m_framebuffer->isValid() || !m_shadersValid) {
-        return;
-    }
-
```
Comment (Holzhaus) [inline src/waveform/renderers/glslwaveformrenderersignal.cpp:253]: Why was this removed? Is it really necessary?

**Review by Holzhaus — COMMENTED** [review]
On AMD/Mesa I get ~33% CPU usage with RGL/GL waveforms at 60fps on AMD/Mesa, both with and without these patches. So no performance regression for me.

File: src/waveform/renderers/glslwaveformrenderersignal.cpp (hunk 4 above)
Comment (Be-ing) [reply src/waveform/renderers/glslwaveformrenderersignal.cpp:253]: This would prevent reaching the `maybeInitializeGL` call below.

Comment (Holzhaus) [reply src/waveform/renderers/glslwaveformrenderersignal.cpp:253]: OK.

**Comment by Holzhaus** [comment]
@uklotzde @daschuer @ronso0 Does anyone of you have an Nvidia GPU? Then it would be great if you could test this. If not, we can merge and hope for the best.

**Comment by ronso0** [comment]
sorry, I have an integrated Intel GPU

**Comment by NotYourAverageAl** [comment]
Ahoy, I have laptop with an Nvidia GPU, what exactly would you like me to test.
So many waveform & spinny PRs around :)

**Comment by Be-ing** [comment]
What OS do you use? The crashing bug only affected macOS. On other OSes we want to confirm that performance is no worse than before.

**Comment by NotYourAverageAl** [comment]
running Windows 10. I'll have a look, and confirm back.

**Comment by NotYourAverageAl** [comment]
No performance regression for me. Tested on two W10 laptops one with integrated gpu & the other an Nvidia gpu

**Comment by Be-ing** [comment]
Okay, I think it's clear this doesn't introduce any performance regressions and it's nice that it benefits at least some Windows users. @daschuer do you have any comments about the code? If not, then I think this is ready for merge.

**Comment by uklotzde** [comment]
@daschuer ?

**Review by uklotzde — APPROVED** [review]

**Comment by daschuer** [comment]
I have not yet compiled the code.
A good test is to use the red/site flicker waveform to compare if it is smooth flickering or not.
I will do this test once I find time to build this.
The code LGTM.

**Comment by Be-ing** [comment]
This has already been tested on 6 different systems. I don't think it needs more testing.

**Pull request merged**